core/services/keeper/registry_synchronizer_sync.go

```diff
@@ -58,20 +67,80 @@ func (rs *RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
 	if nextUpkeepID > countOnContract {
 		return errors.New("invariant, contract should always have at least as many upkeeps as DB")
 	}
+	// newUpkeeps is the range nextUpkeepID, nextUpkeepID + 1 , ... , countOnContract-1
+	newUpkeeps := make([]*big.Int, countOnContract-nextUpkeepID)
+	for i := range newUpkeeps {
+		newUpkeeps[i].Add(big.NewInt(nextUpkeepID), big.NewInt(int64(i)))
+	}
+	rs.batchSyncUpkeepsOnRegistry(reg, newUpkeeps)
 
-	rs.batchSyncUpkeepsOnRegistry(reg, nextUpkeepID, countOnContract)
+	// Delete upkeeps which have been cancelled
+	canceledBigs, err := rs.contract1_1.GetCanceledUpkeepList(nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to get canceled upkeep list")
+	}
+	canceled := make([]int64, len(canceledBigs))
+	for idx, upkeepID := range canceledBigs {
+		canceled[idx] = upkeepID.Int64()
+	}
+	if _, err := rs.orm.BatchDeleteUpkeepsForJob(rs.job.ID, canceled); err != nil {
+		return errors.Wrap(err, "failed to batch delete upkeeps from job")
+	}
+	return nil
+}
+
+func (rs *RegistrySynchronizer) fullSyncUpkeeps1_2(reg Registry) error {
+	startIndex := big.NewInt(0)
+	maxCount := big.NewInt(0)
+	// TODO (sc-37024): Get active upkeep IDs from contract in batches
+	activeUpkeepIDs, err := rs.contract1_2.GetActiveUpkeepIDs(nil, startIndex, maxCount)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get active upkeep IDs")
+	}
+	existingUpkeepIDs, err := rs.orm.AllUpkeepIDsForRegistry(reg.ID)
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch existing upkeep IDs from DB")
+	}
+
+	// New upkeeps are all elements in activeUpkeepIDs which are not in existingUpkeepIDs
+	newUpkeeps := make([]*big.Int, 0)
+	existingSet := make(map[*big.Int]bool)
+	for _, upkeepID := range existingUpkeepIDs {
+		existingSet[upkeepID.ToInt()] = true
+	}
+	for _, upkeepID := range activeUpkeepIDs {
+		if _, found := existingSet[upkeepID]; !found {
+			newUpkeeps = append(newUpkeeps, upkeepID)
+		}
+	}
+	rs.batchSyncUpkeepsOnRegistry(reg, newUpkeeps)
+
+	// All upkeeps in existingUpkeepIDs, not in activeUpkeepIDs should be deleted
+	activeSet := make(map[*big.Int]bool)
+	for _, upkeepID := range activeUpkeepIDs {
+		activeSet[upkeepID] = true
+	}
```

Comment:
here you are looping thru `activeUpkeepIDs` put there is already a loop for that at line 111. we could do this work at the same time i think.